### PR TITLE
Update handling of limited register during consecutive registers allocation

### LINKED ISSUE #84536: SPMI replay failure for Arm64 VectorTableLookup
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=232670
Build error leg or test failing: Build / SuperPMI replay windows x64 checked / Send job to Helix (Windows)
Pull request: https://github.com/dotnet/runtime/pull/84498

## Error Message
Fill the error message using [known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "Assertion failed '!\"Spill candidate has no assignedInterval recentRefPosition\"' in 'JIT.HardwareIntrinsics.Arm._AdvSimd.VectorLookup_3Test__VectorTableLookupByte",
  "BuildRetry": false,
  "ErrorPattern": "",
  "ExcludeConsoleLog": false
}
```

#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|0|

### PR DESCRIPTION
**Pull request by kunalspathak**
Earlier, I added a limit mask to include few registers in case we are allocating during consecutive registers. However, the mask was not sufficient, and we were still running into situation where we won't have register to allocate to one of the consecutive registers or a refPosition that is live at the same location. Updated the handling of such cases:

1. Fix issue where we might have more than one busy register range available to allocate and in prevRegOpt heuristic, teach it to handle the free registers.
2. If we see that consecutive registers are not available in JitStressRegs mode, add back a register range so it can allocate for it.
3. Likewise, if a refposition that is alive at the same location where consecutive registers are live, most of the registers are busy and hence, just do not further limit the registers for such refpositions.
4. While calculating `minRegCount`, take into account the refpositions for consecutive registers.
5. With all of that, we do not need `LsraLimitFPSetForConsecutive`.

Fixes: #84536

### CODE REVIEW

**Review by BruceForstall — COMMENTED** [review]
Couple questions

**Comment by kunalspathak** [comment]
Seems build failures were fixed by https://github.com/dotnet/runtime/pull/84595.

File: src/coreclr/jit/lsra.cpp [hunk 1]
```diff
@@ -12195,6 +12241,10 @@ regMaskTP LinearScan::RegisterSelection::select(Interval*    currentInterval,
         regMaskTP busyRegs = linearScan->regsBusyUntilKill | linearScan->regsInUseThisLocation;
         candidates &= ~busyRegs;
 
+#ifdef DEBUG
+        inUseOrBusyRegsMask |= ~busyRegs;
```
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:12245]: Below, I am using this as `registerAssignment & inUseOrBusyRegsMask`.

File: src/coreclr/jit/lsra.cpp [hunk 2]
```diff
@@ -12210,6 +12260,9 @@ regMaskTP LinearScan::RegisterSelection::select(Interval*    currentInterval,
                 (refPosition->delayRegFree && (checkConflictLocation == (refPosition->nodeLocation + 1))))
             {
                 candidates &= ~checkConflictBit;
+#ifdef DEBUG
+                inUseOrBusyRegsMask |= ~checkConflictBit;
```
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:12264]: Below, I am using this as `registerAssignment & inUseOrBusyRegsMask`.

File: src/coreclr/jit/lsra.cpp [hunk 3]
```diff
@@ -12287,12 +12331,47 @@ regMaskTP LinearScan::RegisterSelection::select(Interval*    currentInterval,
...
+                //                
+                regMaskTP limitCandidatesForConsecutive = refPosition->registerAssignment & inUseOrBusyRegsMask;
+                regMaskTP overallLimitCandidates;
+                regMaskTP limitConsecutiveResult =
+                    linearScan->filterConsecutiveCandidates(limitCandidatesForConsecutive, refPosition->regCount,
+                                                            &overallLimitCandidates);
+                assert(limitConsecutiveResult != RBM_NONE);
+
+                DWORD startRegister = 0;
+                BitScanForward64(&startRegister, static_cast<DWORD64>(limitConsecutiveResult));
```
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:12352]: agree. I thought I was using it at one point, but then somehow ended up using the macro.

File: src/coreclr/jit/lsra.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:12264]: That's true.

**Comment by tannergooding** [comment]
> superpmi errors are https://github.com/dotnet/runtime/issues/84536

Isn't this PR supposed to fix this failure in particular?

**Comment by kunalspathak** [comment]
> > superpmi errors are #84536
> 
> Isn't this PR supposed to fix this failure in particular?

Ah, meant to have this comment for the other PR. Will update my comment.

**Review by BruceForstall — APPROVED** [review]

**Comment by tannergooding** [comment]
-- As with https://github.com/dotnet/runtime/pull/84634#issuecomment-1504240772, it looks like the SPMI asserts that are still present are from `\base\`.

This PR, compared to a previous PR has "half" the asserts with older PRs listing failures for both `\base\` and `\diff\`.

**Review by tannergooding — APPROVED** [review]

**Pull request merged**